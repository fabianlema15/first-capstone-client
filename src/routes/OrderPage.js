import React from 'react';
import GenericContext from '../contexts/GenericContext'
import OrderContent from '../components/Order/Order'
import GenericApiService from '../services/generic-api-service';
import Helper from '../components/Utils/Helper';
import TokenService from '../services/token-service'

class OrderPage extends React.Component{
  static defaultProps = {
    ROUTE : 'orders'
  }

  state = {
    objList: {},
    error: null,
    showForm: false,
  };

  setOrderList = objList => {
    this.setState({ objList })
  }

  setError = error => {
    this.setState({ error: error.error })
  }

  clearError = () => {
    this.setState({ error: null })
  }

  showHideForm = show => {
    this.setState({ showForm: show })
  }

  setSelected = (idOrder, selected=true) => {
    this.props.history.push(`/order/${idOrder}`)
    //return <Redirect to={`/orders/${idOrder}`} />
  }

  addOrder = order => {
    this.setState({
      objList: {
        ...this.state.objList,
        [order.id]:order},
      showForm: false
    })
  }

  editOrder = (idOrder, orderUpdated) => {
    orderUpdated.selected = false
    Object.assign(this.state.objList[idOrder], orderUpdated);
    this.setState({
      objList:{
        ...this.state.objList,
        [idOrder]:this.state.objList[idOrder]}
    })
  }

  removeOrder = (idOrder) => {
    const { [idOrder]: id, ...rest } = this.state.objList;
    this.setState({
      objList: rest
    })
  }

  getAll = () => {
    if (TokenService.decodeUser())
      GenericApiService.getAll(`${this.props.ROUTE}/byuser/${TokenService.decodeUser().user_id}`)
        .then(result => this.setOrderList(Helper.serializeObj(result)))
        .catch(this.setError)
  }

  submitNew = (e) => {
    e.preventDefault();
    const { name, picture, stock, price, description } = e.target;
    const newOrder = {
      name: name.value,
      picture: picture.value,
      stock: stock.value,
      price: price.value,
      description: description.value
    }

    GenericApiService.saveNew(this.props.ROUTE, newOrder)
      .then(this.addOrder)
      .catch(this.setError)
  }

  submitEdit = (e) => {
    e.preventDefault();
    const { id, name, picture, stock, price, description } = e.target;
    const newOrder = {
      name: name.value,
      picture: picture.value,
      stock: stock.value,
      price: price.value,
      description: description.value
    }

    GenericApiService.saveExisting(this.props.ROUTE, id.value, newOrder)
      .then(() => this.editOrder(id.value, newOrder))
      .catch(this.setError)
  }

  delete = (idOrdero) => {
    GenericApiService.delete(this.props.ROUTE, idOrdero)
      .then(() => this.removeOrder(idOrdero))
      .catch(this.setError)
  }

  render() {
    const value = {
      objList: this.state.objList,
      showForm: this.state.showForm,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      showHideForm: this.showHideForm,
      setSelected: this.setSelected,
      getAll: this.getAll,
      submitNew: this.submitNew,
      submitEdit: this.submitEdit,
      delete: this.delete,
      getObjArray: Helper.getOrdersArray(),
      objName: 'Order'
    }

    return (
      <GenericContext.Provider value={value}>
        <OrderContent />
      </GenericContext.Provider>
    )
  }
}

export default OrderPage;
