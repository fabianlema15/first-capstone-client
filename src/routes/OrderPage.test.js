import React from 'react';
import ReactDOM from 'react-dom';
import OrderPage from './OrderPage'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer';

describe('<OrderPage />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
      <OrderPage>
      </OrderPage>
    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Renders OrderPage as expected', () => {
	    const tree = renderer.create(<BrowserRouter>
        <OrderPage>
        </OrderPage>
      </BrowserRouter>).toJSON();
	    expect(tree).toMatchSnapshot();
	  });

})
