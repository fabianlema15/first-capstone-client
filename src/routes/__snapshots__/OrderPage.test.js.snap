// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderPage /> Renders OrderPage as expected 1`] = `
<div>
  <nav>
    <div
      className="back-nav"
    >
      <a
        href="/menu"
        onClick={[Function]}
      >
        Back
      </a>
    </div>
    <div
      className="name-nav"
    >
      <h2>
        Order
        s
      </h2>
    </div>
  </nav>
  <main
    className="content-with-nav"
  >
    <section>
      <a
        href="/orders/new"
        onClick={[Function]}
      >
        Create New Order
      </a>
    </section>
    <section>
      <div>
        You still have no Orders added
      </div>
    </section>
  </main>
</div>
`;
